fault_injection/fi_ibex.py

```diff
@@ -0,0 +1,200 @@
+    # target_clk_mult is a hardcoded constant in the FPGA bitstream.
+    cfg["target"]["pll_frequency"] = cfg["target"]["target_freq"] / cfg["target"]["target_clk_mult"]
+
+    # Init target.
+    logger.info(f"Initializing target {cfg['target']['target_type']} ...")
+    target = CWFPGA(
+        bitstream = cfg["target"]["fpga_bitstream"],
+        force_programming = cfg["target"]["force_program_bitstream"],
+        firmware = cfg["target"]["fw_bin"],
+        pll_frequency = cfg["target"]["pll_frequency"],
+        baudrate = cfg["target"]["baudrate"],
+        output_len = cfg["target"]["output_len_bytes"],
+        protocol = cfg["target"]["protocol"]
+    )
+
+    # Init FI gear.
+    fi_gear = FIGear(cfg)
+
+    # Init project.
+    project_cfg = ProjectConfig(type = cfg["fiproject"]["project_db"],
+                                path = project,
+                                overwrite = True,
+                                fi_threshold = cfg["fiproject"].get("project_mem_threshold")
+                                )
+    project = FIProject(project_cfg)
+    project.create_project()
+
+    return target, fi_gear, project
+
+
+def fi_parameter_sweep(cfg: dict, cwtarget: CWFPGA, fi_gear,
+                       project: FIProject, ot_communication: OTFIIbex) -> None:
+    """ Fault parameter sweep.
+
+    Sweep through the fault parameter space.
+
+    Args:
+        cfg: The FI project configuration.
+        cwtarget: The OpenTitan target.
+        fi_gear: The FI gear to use.
+        project: The project to store the results.
+        ot_communication: The OpenTitan Ibex FI communication interface.
+    """
+    # Configure the trigger.
+    ot_communication.init_trigger()
+    # Start the parameter sweep.
+    remaining_iterations = cfg["fisetup"]["num_iterations"]
+    with tqdm(total=remaining_iterations, desc="Injecting", ncols=80,
+              unit=" different faults") as pbar:
+        while remaining_iterations > 0:
+            # Get fault parameters (e.g., trigger delay, glitch voltage).
+            fault_parameters = fi_gear.generate_fi_parameters()
+
+            # Arm the FI gear.
+            fi_gear.arm_trigger(fault_parameters)
+
+            # Start test on OpenTitan.
+            ot_communication.start_test(cfg)
+
+            # Read response.
+            response = ot_communication.read_response()
+
+            # Compare response.
+            # Check if result is expected result (FI failed), unexpected result
+            # (FI successful), or no response (FI failed.)
+            fi_result = FISuccess.SUCCESS
+            if response == cfg["test"]["expected_result"]:
+                # Expected result received. No FI effect.
+                fi_result = FISuccess.EXPRESPONSE
+            elif response == "":
+                # No UART response received.
+                fi_result = FISuccess.NORESPONSE
+                # Resetting OT as it most likely crashed.
+                cwtarget.reset_target()
```

Comment:
Would it make sense to first only try resetting OT instead of directly re-flashing the bitstream?